// electron-forge/dist/api/make.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _bluebird = require('bluebird');

require('colors');

var _fsPromise = require('fs-promise');

var _fsPromise2 = _interopRequireDefault(_fsPromise);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _oraHandler = require('../util/ora-handler');

var _oraHandler2 = _interopRequireDefault(_oraHandler);

var _electronHostArch = require('../util/electron-host-arch');

var _electronHostArch2 = _interopRequireDefault(_electronHostArch);

var _forgeConfig = require('../util/forge-config');

var _forgeConfig2 = _interopRequireDefault(_forgeConfig);

var _messages = require('../util/messages');

var _readPackageJson = require('../util/read-package-json');

var _readPackageJson2 = _interopRequireDefault(_readPackageJson);

var _requireSearch = require('../util/require-search');

var _requireSearch2 = _interopRequireDefault(_requireSearch);

var _resolveDir = require('../util/resolve-dir');

var _resolveDir2 = _interopRequireDefault(_resolveDir);

var _package = require('./package');

var _package2 = _interopRequireDefault(_package);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @typedef {Object} MakeOptions
 * @property {string} [dir=process.cwd()] The path to the app from which distributables are generated
 * @property {boolean} [interactive=false] Whether to use sensible defaults or prompt the user visually
 * @property {boolean} [skipPackage=false] Whether to skip the pre-make packaging step
 * @property {Array<string>} [overrideTargets] An array of make targets to override your forge config
 * @property {string} [arch=host architecture] The target architecture
 * @property {string} [platform=process.platform] The target platform. NOTE: This is limited to be the current platform at the moment
 */

/**
 * Make distributables for an Electron application.
 *
 * @param {MakeOptions} providedOptions - Options for the make method
 * @return {Promise} Will resolve when the make process is complete
 */
exports.default = function () {
  var _ref = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee3() {
    var providedOptions = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var _Object$assign, dir, interactive, skipPackage, overrideTargets, arch, platform, forgeConfig, declaredArch, declaredPlatform, targets, targetArchs, packageJSON, appName, outputs, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _loop, _iterator, _step;

    return _regenerator2.default.wrap(function _callee3$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            // eslint-disable-next-line prefer-const, no-unused-vars
            _Object$assign = (0, _assign2.default)({
              dir: process.cwd(),
              interactive: false,
              skipPackage: false,
              arch: (0, _electronHostArch2.default)(),
              platform: process.platform
            }, providedOptions), dir = _Object$assign.dir, interactive = _Object$assign.interactive, skipPackage = _Object$assign.skipPackage, overrideTargets = _Object$assign.overrideTargets, arch = _Object$assign.arch, platform = _Object$assign.platform;

            _oraHandler2.default.interactive = interactive;

            forgeConfig = void 0;
            _context5.next = 5;
            return (0, _oraHandler2.default)('Resolving Forge Config', (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee() {
              return _regenerator2.default.wrap(function _callee$(_context) {
                while (1) {
                  switch (_context.prev = _context.next) {
                    case 0:
                      _context.next = 2;
                      return (0, _resolveDir2.default)(dir);

                    case 2:
                      dir = _context.sent;

                      if (dir) {
                        _context.next = 5;
                        break;
                      }

                      throw 'Failed to locate makeable Electron application';

                    case 5:
                      _context.next = 7;
                      return (0, _forgeConfig2.default)(dir);

                    case 7:
                      forgeConfig = _context.sent;

                    case 8:
                    case 'end':
                      return _context.stop();
                  }
                }
              }, _callee, undefined);
            })));

          case 5:
            if (!(platform && platform !== process.platform && !(process.platform === 'darwin' && platform === 'mas'))) {
              _context5.next = 7;
              break;
            }

            throw "You cannot run 'make' for a platform other than your system's platform";

          case 7:
            if (skipPackage) {
              _context5.next = 13;
              break;
            }

            (0, _messages.info)(interactive, 'We need to package your application before we can make it'.green);
            _context5.next = 11;
            return (0, _package2.default)({
              dir: dir,
              interactive: interactive,
              arch: arch,
              platform: platform
            });

          case 11:
            _context5.next = 14;
            break;

          case 13:
            (0, _messages.warn)(interactive, 'WARNING: Skipping the packaging step, this could result in an out of date build'.red);

          case 14:
            declaredArch = arch;
            declaredPlatform = platform;
            targets = forgeConfig.make_targets[declaredPlatform];

            if (overrideTargets) {
              targets = overrideTargets;
            }

            (0, _messages.info)(interactive, 'Making for the following targets:', ('' + targets.join(', ')).cyan);

            targetArchs = [declaredArch];

            if (!(declaredArch === 'all')) {
              _context5.next = 30;
              break;
            }

            _context5.t0 = process.platform;
            _context5.next = _context5.t0 === 'darwin' ? 24 : _context5.t0 === 'linux' ? 26 : _context5.t0 === 'win32' ? 28 : 28;
            break;

          case 24:
            targetArchs = ['x64'];
            return _context5.abrupt('break', 30);

          case 26:
            targetArchs = ['ia32', 'x64', 'armv7l'];
            return _context5.abrupt('break', 30);

          case 28:
            targetArchs = ['ia32', 'x64'];
            return _context5.abrupt('break', 30);

          case 30:
            _context5.next = 32;
            return (0, _readPackageJson2.default)(dir);

          case 32:
            packageJSON = _context5.sent;
            appName = forgeConfig.electronPackagerConfig.name || packageJSON.productName || packageJSON.name;
            outputs = [];
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            _context5.prev = 38;
            _loop = _regenerator2.default.mark(function _loop() {
              var targetArch, packageDir, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _loop2, _iterator2, _step2;

              return _regenerator2.default.wrap(function _loop$(_context4) {
                while (1) {
                  switch (_context4.prev = _context4.next) {
                    case 0:
                      targetArch = _step.value;
                      packageDir = _path2.default.resolve(dir, 'out/' + appName + '-' + declaredPlatform + '-' + targetArch);
                      _context4.next = 4;
                      return _fsPromise2.default.exists(packageDir);

                    case 4:
                      if (_context4.sent) {
                        _context4.next = 6;
                        break;
                      }

                      throw new Error('Couldn\'t find packaged app at: ' + packageDir);

                    case 6:
                      _iteratorNormalCompletion2 = true;
                      _didIteratorError2 = false;
                      _iteratorError2 = undefined;
                      _context4.prev = 9;
                      _loop2 = _regenerator2.default.mark(function _loop2() {
                        var target;
                        return _regenerator2.default.wrap(function _loop2$(_context3) {
                          while (1) {
                            switch (_context3.prev = _context3.next) {
                              case 0:
                                target = _step2.value;
                                _context3.next = 3;
                                return (0, _oraHandler2.default)('Making for target: ' + target.cyan + ' - On platform: ' + declaredPlatform.cyan + ' - For arch: ' + targetArch.cyan, (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee2() {
                                  var maker;
                                  return _regenerator2.default.wrap(function _callee2$(_context2) {
                                    while (1) {
                                      switch (_context2.prev = _context2.next) {
                                        case 0:
                                          maker = (0, _requireSearch2.default)(__dirname, ['../makers/' + process.platform + '/' + target + '.js', '../makers/generic/' + target + '.js', 'electron-forge-maker-' + target]);

                                          if (maker) {
                                            _context2.next = 3;
                                            break;
                                          }

                                          throw 'Could not find a build target with the name: ' + target + ' for the platform: ' + declaredPlatform;

                                        case 3:
                                          _context2.prev = 3;
                                          _context2.t0 = outputs;
                                          _context2.next = 7;
                                          return (maker.default || maker)(packageDir, appName, targetArch, forgeConfig, packageJSON);

                                        case 7:
                                          _context2.t1 = _context2.sent;

                                          _context2.t0.push.call(_context2.t0, _context2.t1);

                                          _context2.next = 18;
                                          break;

                                        case 11:
                                          _context2.prev = 11;
                                          _context2.t2 = _context2['catch'](3);

                                          if (!_context2.t2) {
                                            _context2.next = 17;
                                            break;
                                          }

                                          throw {
                                            message: 'An error occured while making for target: ' + target,
                                            stack: _context2.t2.message + '\n' + _context2.t2.stack
                                          };

                                        case 17:
                                          throw new Error('An unknown error occured while making for target: ' + target);

                                        case 18:
                                        case 'end':
                                          return _context2.stop();
                                      }
                                    }
                                  }, _callee2, undefined, [[3, 11]]);
                                })));

                              case 3:
                              case 'end':
                                return _context3.stop();
                            }
                          }
                        }, _loop2, undefined);
                      });
                      _iterator2 = (0, _getIterator3.default)(targets);

                    case 12:
                      if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
                        _context4.next = 17;
                        break;
                      }

                      return _context4.delegateYield(_loop2(), 't0', 14);

                    case 14:
                      _iteratorNormalCompletion2 = true;
                      _context4.next = 12;
                      break;

                    case 17:
                      _context4.next = 23;
                      break;

                    case 19:
                      _context4.prev = 19;
                      _context4.t1 = _context4['catch'](9);
                      _didIteratorError2 = true;
                      _iteratorError2 = _context4.t1;

                    case 23:
                      _context4.prev = 23;
                      _context4.prev = 24;

                      if (!_iteratorNormalCompletion2 && _iterator2.return) {
                        _iterator2.return();
                      }

                    case 26:
                      _context4.prev = 26;

                      if (!_didIteratorError2) {
                        _context4.next = 29;
                        break;
                      }

                      throw _iteratorError2;

                    case 29:
                      return _context4.finish(26);

                    case 30:
                      return _context4.finish(23);

                    case 31:
                    case 'end':
                      return _context4.stop();
                  }
                }
              }, _loop, undefined, [[9, 19, 23, 31], [24,, 26, 30]]);
            });
            _iterator = (0, _getIterator3.default)(targetArchs);

          case 41:
            if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
              _context5.next = 46;
              break;
            }

            return _context5.delegateYield(_loop(), 't1', 43);

          case 43:
            _iteratorNormalCompletion = true;
            _context5.next = 41;
            break;

          case 46:
            _context5.next = 52;
            break;

          case 48:
            _context5.prev = 48;
            _context5.t2 = _context5['catch'](38);
            _didIteratorError = true;
            _iteratorError = _context5.t2;

          case 52:
            _context5.prev = 52;
            _context5.prev = 53;

            if (!_iteratorNormalCompletion && _iterator.return) {
              _iterator.return();
            }

          case 55:
            _context5.prev = 55;

            if (!_didIteratorError) {
              _context5.next = 58;
              break;
            }

            throw _iteratorError;

          case 58:
            return _context5.finish(55);

          case 59:
            return _context5.finish(52);

          case 60:
            return _context5.abrupt('return', outputs);

          case 61:
          case 'end':
            return _context5.stop();
        }
      }
    }, _callee3, undefined, [[38, 48, 52, 60], [53,, 55, 59]]);
  }));

  return function (_x) {
    return _ref.apply(this, arguments);
  };
}();
