// electron-forge/dist/makers/win32/appx.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDefaultCertificate = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _bluebird = require('bluebird');

var createDefaultCertificate = exports.createDefaultCertificate = function () {
  var _ref = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee(publisherName, outPath) {
    var defaultPvk, targetCert, targetPfx;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            defaultPvk = _path2.default.resolve(__dirname, '..', '..', '..', 'res', 'default.pvk');
            targetCert = _path2.default.join(outPath, 'default.cer');
            targetPfx = _path2.default.join(outPath, 'default.pfx');
            _context.next = 5;
            return spawnSdkTool('makecert.exe', ['-r', '-h', '0', '-n', 'CN=' + publisherName, '-eku', '1.3.6.1.5.5.7.3.3', '-pe', '-sv', defaultPvk, targetCert]);

          case 5:
            _context.next = 7;
            return spawnSdkTool('pvk2pfx.exe', ['-pvk', defaultPvk, '-spc', targetCert, '-pfx', targetPfx]);

          case 7:
            return _context.abrupt('return', targetPfx);

          case 8:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function createDefaultCertificate(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

var _electronWindowsStore = require('electron-windows-store');

var _electronWindowsStore2 = _interopRequireDefault(_electronWindowsStore);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _spawnRx = require('spawn-rx');

var _ensureOutput = require('../../util/ensure-output');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// NB: This is not a typo, we require AppXs to be built on 64-bit
// but if we're running in a 32-bit node.js process, we're going to
// be Wow64 redirected
var windowsSdkPath = process.arch === 'x64' ? 'C:\\Program Files (x86)\\Windows Kits\\10\\bin\\x64' : 'C:\\Program Files\\Windows Kits\\10\\bin\\x64';

function findSdkTool(exe) {
  var sdkTool = _path2.default.join(windowsSdkPath, exe);
  if (!_fs2.default.existsSync(sdkTool)) {
    sdkTool = (0, _spawnRx.findActualExecutable)(exe, []).cmd;
  }

  if (!_fs2.default.existsSync(sdkTool)) {
    throw new Error('Can\'t find ' + exe + ' in PATH, you probably need to install the Windows SDK');
  }

  return sdkTool;
}

function spawnSdkTool(exe, params) {
  return (0, _spawnRx.spawnPromise)(findSdkTool(exe), params);
}

exports.default = function () {
  var _ref2 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee2(dir, appName, targetArch, forgeConfig, packageJSON) {
    var outPath, opts, noBeta, err;
    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            // eslint-disable-line
            outPath = _path2.default.resolve(dir, '../make/appx/' + targetArch);
            _context2.next = 3;
            return (0, _ensureOutput.ensureDirectory)(outPath);

          case 3:
            opts = (0, _assign2.default)({}, forgeConfig.windowsStoreConfig, {
              inputDirectory: dir,
              outputDirectory: outPath,
              publisher: packageJSON.author,
              flatten: false,
              deploy: false,
              packageVersion: packageJSON.version + '.0',
              packageName: appName.replace(/-/g, ''),
              packageDisplayName: appName,
              packageDescription: packageJSON.description || appName,
              packageExecutable: 'app\\' + appName + '.exe',
              windowsKit: _path2.default.dirname(findSdkTool('makeappx.exe'))
            });

            if (opts.devCert) {
              _context2.next = 8;
              break;
            }

            _context2.next = 7;
            return createDefaultCertificate(opts.publisher, outPath);

          case 7:
            opts.devCert = _context2.sent;

          case 8:

            if (!opts.publisher.match(/^CN=/)) {
              opts.publisher = 'CN=' + opts.publisher;
            }

            if (!opts.packageVersion.match(/-/)) {
              _context2.next = 17;
              break;
            }

            if (!opts.makeVersionWinStoreCompatible) {
              _context2.next = 15;
              break;
            }

            noBeta = opts.packageVersion.replace(/-.*/, '');

            opts.packageVersion = noBeta + '.0';
            _context2.next = 17;
            break;

          case 15:
            err = "Windows Store version numbers don't support semver beta tags. To" + 'automatically fix this, set makeVersionWinStoreCompatible to true or ' + 'explicitly set packageVersion to a version of the format X.Y.Z.A';
            throw new Error(err);

          case 17:

            delete opts.makeVersionWinStoreCompatible;

            _context2.next = 20;
            return (0, _electronWindowsStore2.default)(opts);

          case 20:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, undefined);
  }));

  return function (_x3, _x4, _x5, _x6, _x7) {
    return _ref2.apply(this, arguments);
  };
}();
